### PR TITLE
Let each nonlinear system select its own solver

### PR DESCRIPTION
**Pull request by phannebohm**
By default each system gets the globally selected solver. If the system is sparse, kinsol is selected.
Also do a more compact dump.

### Related Issues

issue #7848

### CODE REVIEW

**Comment by casella** [comment]
@phannebohm, can you briefly summarize the current status after your last commit?

**Comment by phannebohm** [comment]
First of all, two blunders on my account:
1. after some detours I finally found the official documentation for kinsol ([for the record](https://raw.githubusercontent.com/LLNL/sundials/master/doc/kinsol/kin_guide.pdf)).
2. after debugging far too long, I just tried running the failing models and am unable to reproduce the issue...
I even found some other _failing tests_ that work, at least on my machine.

Anyhow, the Jenkins job does some weird stuff and won't even pass setup which should be beyond this PR's influence. So I force pushed to restart the job, sadly with the same result as before. At this point I give up for the moment and come back when Jenkins has sorted out its issues.

**Comment by perost** [comment]
The Jenkins issue should already be fixed since a few hours back.

**Comment by casella** [comment]
Cross fingers :)

**Comment by adrpo** [comment]
I ran one of the bad tests several times on the same machine, 2 times bad, 1 time ok, seems random:

```
hudson@ripper2:/var/lib/jenkins2/ws/OpenModelica_PR-7933/testsuite/simulation/modelica/nonlinear_system$ ./nonlinear_system.problem6
stdout            | info    | Using sparse solver kinsol for nonlinear system 0 (198),
|                 | |       | because density of 0.03 remains under threshold of 0.20.
stdout            | info    | Using sparse solver kinsol for nonlinear system 1 (395),
|                 | |       | because density of 0.03 remains under threshold of 0.20.
stdout            | info    | The maximum density for using sparse solvers can be specified
|                 | |       | using the runtime flag '<-nlsMaxDensity=value>'.
assert            | debug   | ##SUNLS## In function SUNLinSol_KLUReInit: Illegal function input.
assert            | debug   | Solving non-linear system 198 failed at time=0.
|                 | |       | For more information please use -lv LOG_NLS.
assert            | info    | simulation terminated by an assertion at initialization
hudson@ripper2:/var/lib/jenkins2/ws/OpenModelica_PR-7933/testsuite/simulation/modelica/nonlinear_system$ ./nonlinear_system.problem6
stdout            | info    | Using sparse solver kinsol for nonlinear system 0 (198),
|                 | |       | because density of 0.03 remains under threshold of 0.20.
stdout            | info    | Using sparse solver kinsol for nonlinear system 1 (395),
|                 | |       | because density of 0.03 remains under threshold of 0.20.
stdout            | info    | The maximum density for using sparse solvers can be specified
|                 | |       | using the runtime flag '<-nlsMaxDensity=value>'.
assert            | debug   | ##SUNLS## In function SUNLinSol_KLUReInit: Illegal function input.
assert            | debug   | Solving non-linear system 198 failed at time=0.
|                 | |       | For more information please use -lv LOG_NLS.
assert            | info    | simulation terminated by an assertion at initialization
hudson@ripper2:/var/lib/jenkins2/ws/OpenModelica_PR-7933/testsuite/simulation/modelica/nonlinear_system$ ./nonlinear_system.problem6
stdout            | info    | Using sparse solver kinsol for nonlinear system 0 (198),
|                 | |       | because density of 0.03 remains under threshold of 0.20.
stdout            | info    | Using sparse solver kinsol for nonlinear system 1 (395),
|                 | |       | because density of 0.03 remains under threshold of 0.20.
stdout            | info    | The maximum density for using sparse solvers can be specified
|                 | |       | using the runtime flag '<-nlsMaxDensity=value>'.
LOG_SUCCESS       | info    | The initialization finished successfully without homotopy method.
LOG_SUCCESS       | info    | The simulation finished successfully.

```

will try valgrind on that now.

**Comment by casella** [comment]
Almost three, only [3 tests failing](https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-7933/19/tests)...

**Comment by perost** [comment]
> Almost three, only [3 tests failing](https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-7933/19/tests)...

Unfortunately only with the gcc testing and not clang, normally the same tests would be expected to fail with both. So it looks like there's still some undefined behaviour due to memory issues.

**Comment by casella** [comment]
> 
> 
> > Almost three, only [3 tests failing](https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-7933/19/tests)...
> 
> Unfortunately only with the gcc testing and not clang, normally the same tests would be expected to fail with both. So it looks like there's still some undefined behaviour due to memory issues.

Argh, you're right. Good luck with that...

**Comment by phannebohm** [comment]
> Almost three, only [3 tests failing](https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-7933/19/tests)...

Actually it would have been only two. The third one was previously part of the failing tests but for clang it works so I was bold.

> > Unfortunately only with the gcc testing and not clang, normally the same tests would be expected to fail with both. So it looks like there's still some undefined behaviour due to memory issues.
> 
> Argh, you're right. In fact "assert | debug | ##SUNLS## In function SUNLinSol_KLUReInit: Illegal function input." sounded a bit suspect to me. @phannebohm, maybe you can just find out with a plain C debugger.
> 
> Good luck with that...

There is in fact nothing I can do, at least on my local machine, because here it all works so there is nothing to debug...
I would need to debug on the server but I have no idea how to do that. Any suggestions? I remember @AnHeuermann suggesting docker images but I have no experience with that.

The reason I updated this PR was because I was hoping the problem disappeared by itself in the meantime :sweat_smile:

**Comment by phannebohm** [comment]
> Almost three, only [3 tests failing](https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-7933/19/tests)...

I could just mark those 3 as failing so the PR gets through. I mean there are a few people waiting for those changes right?
I would assume the kinsol issue has been there anyway, my changes just expose it (for gcc only).
If I can get a test that also fails on my machine I can start debugging, before that this is my best offer.
Opinions?

**Comment by AnHeuermann** [comment]
I can review you code tomorrow, but i would guess that the kinsol error is new and we should fix it.

Running all in the same image as on Jenkins should be possible. But debugging inside Docker containers is annoying.

**Comment by casella** [comment]
I'm in favour of some more (static) debugging by @AnHeuermann before @phannebohm gives up. The fact that some code fails on some machine, but not others, is a tell-tale sign that it is fragile; e.g., it uses uninitialized variables in some place. I guess we should at least attempt to figure out why.

@adrpo, do you have any suggestions on how could we debug the Jenkins run, or at least try to reproduce its failures on another machine?

File: OMCompiler/SimulationRuntime/c/simulation_data.h [hunk 1]
```diff
@@ -294,7 +294,9 @@ typedef struct NONLINEAR_SYSTEM_DATA
   void (*getIterationVars)(struct DATA*, double*);
   int (*checkConstraints)(struct DATA*, threadData_t *threadData);
 
+  NONLINEAR_SOLVER method;
```
Comment (AnHeuermann) [inline OMCompiler/SimulationRuntime/c/simulation_data.h:297]: Add some description comment to this.

**Review by AnHeuermann — CHANGES_REQUESTED** [review]
I can't see anything in your commit that could cause this issue, but we should investigate this error.

**Comment by phannebohm** [comment]
I'm not much closer to the root cause yet. If the tests pass, it's purely by chance.

In the random cases when the tests fail the call to `KINSol` at
https://github.com/phannebohm/OpenModelica/blob/321e8b5c9e447c324ee344f8b2a6552261d0ce30/OMCompiler/SimulationRuntime/c/simulation/solver/kinsolSolver.c#L1205
returns with message `KIN_LSOLVE_FAIL` meaning the linear system could not be solved I guess.
When KLU is used a reinit for the linear solver is done at this point. I'm not sure what to do for the other nls linerar solvers.
But since this occurs randomly there must be something else going wrong.

**Comment by phannebohm** [comment]
Ok I hope I got it now!

> I'm not sure what to do for the other nls linerar solvers.

Just do nothing and let kinsol retry :smile:

**Comment by phannebohm** [comment]
This didn't work as planned :(
The [regressions](https://libraries.openmodelica.org/branches/history/master/2021-11-12%2022:41:02..2021-11-13%2019:36:18.html) suggest I broke more than I fixed with this. Including all of OpenIPSL. :see_no_evil:

Maybe kinsol needs KLU as its linear solver. I'll try giving each nonlinear system not only a NLS method but also a NLS_LS method.
Another possibility is that the other linear solvers need some kind of reinit, but I could not find that in the [manual](http://runge.math.smu.edu/arkode_dev/doc/guide/build/html/sunlinsol/).

**Comment by casella** [comment]
The linear solver is not the culprit. Besides, we're already using KLU by default with kinsol, see [-nlsLS](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlsls).

The problem is that kinsol is less robust than our default hybrid solver, that also includes homotopy approaches and other black magic, when starting from poor initial guesses. This can be a problem during initialization, and possibly after events causing large discontinuities in the system.

My suggestion is to use this strategy:
- if the system has a density below the [-nlssMaxDensity](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssmaxdensity) limit or a size above the [-nlssMinSize](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssminsize) limit, then try the sparse solver first
- if the sparse solver is used, and it fails, and if the size is below the -nlssMinSize limit, then also try the default dense solver, before giving up
- reduce the [-nlssMinSize](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssminsize) to 1000, from the current value of 10001, which is way too large. 

A system with 10000 equations has 100 million dense jacobian entries, which occupy almost 1 GB of RAM, and it just takes forever to be factorized. The solver may be successful, but after waiting for a few minutes, you would probably think the solver hanged, and just kill the process. 

Based on my experience in the last few years, I would say that 1000 is the absolute maximum limit for initialization. 3000 is already way too big; if you consider that the time to solve sparse systems is approximately linear with the size for sparse solvers, and depends on the cube of size for dense solvers, going from 1000 to 3000 increases the solution time by one order of magnitude.

@phannebohm, can you implement these small improvements? I guess this should be enough to get rid of all those regressions, while guaranteeing fast performance in PowerGrids, which was my original goal.

**Comment by phannebohm** [comment]
> The linear solver is not the culprit. Besides, we're already using KLU by default with kinsol, see [-nlsLS](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlsls). Or at least, this was what I understood from Willi back then.

This is absolutely the case if the user sets kinsol as the solver (for all systems). But with this PR each system selects its own nonlinear solver (choosing between kinsol and the one set by the user), but not yet its own linear solver. At least I'm relatively certain from what I have seen in the code. (However, it is humanly impossible to know 100% what actually happens in the runtime so maybe I'm wrong :sweat_smile:)

> The problem is that kinsol is less robust than our default hybrid solver, that also includes homotopy approaches and other black magic, when starting from poor initial guesses. This can be a problem during initialization, and possibly after events causing large discontinuities in the system.
> 
> My suggestion is to use this strategy:
> 
>     * if the system has a density below the [-nlssMaxDensity](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssmaxdensity) limit or a size above the [-nlssMinSize](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssminsize) limit, then try the sparse solver first
> 
>     * if the sparse solver is used, and it fails, and if the size is below the [-nlssMinSize](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssminsize) limit, then also try the default dense solver, before giving up
> 
>     * reduce the [-nlssMinSize](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssminsize) default to 1000, from the current value of 10001, which is way too large.
> 
> 
> A system with 10000 equations has 100 million dense jacobian entries, which occupy almost 1 GB of RAM, and it just takes forever to be factorized. The solver may be successful, but after waiting for a few minutes, you would probably think the solver hanged, and just kill the process.
> 
> Based on my experience in the last few years, I would say that 1000 is the absolute maximum limit for initialization. 3000 is already way too big; if you consider that the time to solve sparse systems is approximately linear with the size for sparse solvers, and depends on the cube of size for dense solvers, going from 1000 to 3000 increases the ratio between the dense solver time and the sparse solver time by one order of magnitude.
> 
> @phannebohm, can you implement these small improvements? I guess this should be enough to get rid of all those regressions, while guaranteeing fast performance in PowerGrids, which was my original goal.

Thanks for the suggestions, your confidence makes me despair less.
* The first bullet point should already be in place with this PR, right?
* As for the second, I'm not sure how easy it will be to have the system try different solvers in a row. But that should be the way to go.
* The third one should be done after the above is working I guess.

The upcoming week is already full with other stuff, so I can't promise to do it quickly.

**Comment by casella** [comment]
> > The linear solver is not the culprit. Besides, we're already using KLU by default with kinsol, see [-nlsLS](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlsls). Or at least, this was what I understood from Willi back then.
> 
> This is absolutely the case if the user sets kinsol as the solver (for all systems). But with this PR each system selects its own nonlinear solver (choosing between kinsol and the one set by the user), but not yet its own linear solver.

Ah ok, then what you wrote about also setting nlsLS was correct. Yeah, this is absolutely necessary, because the dense solver scales very badly with size.

> > My suggestion is to use this strategy:
> > ```
> > * if the system has a density below the [-nlssMaxDensity](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssmaxdensity) limit or a size above the [-nlssMinSize](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssminsize) limit, then try the sparse solver first
> > 
> > * if the sparse solver is used, and it fails, and if the size is below the [-nlssMinSize](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssminsize) limit, then also try the default dense solver, before giving up
> > 
> > * reduce the [-nlssMinSize](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-nlssminsize) default to 1000, from the current value of 10001, which is way too large.
>     * The first bullet point should already be in place with this PR, right?
Yes.
>     * As for the second, I'm not sure how easy it will be to have the system try different solvers in a row. But that should be the way to go.
Agreed. The principle is: try as many solver or solution strategies as possible, that could help, to maximize the likelyhood of success. It already worked fine in other cases.
> 
>     * ~The third one should be done after the above is working I guess.~ Actually, reducing -nlssMinSize would be independent of the rest, I can do that tmrw.

Please do. 10000 equations with the dense solver makes very little sense.

We should probably also put 1000 as the -lssMinSize default value, for the same reason. Currently, it is set at 4000.

> The upcoming week is already full with other stuff, so I can't promise to do it quickly.

No problem. I'll report these conclusions in #7848 so we don't forget.

**Pull request merged**